File: 0054-spiral-matrix/0054-spiral-matrix.scala
import scala.collection.mutable.ListBuffer

object Solution {
  def spiralOrder(matrix: Array[Array[Int]]): List[Int] = {
    val ans = ListBuffer[Int]()
    var top = 0
    var left = 0
    var bottom = matrix.length - 1
    var right = matrix(0).length - 1
    var direction = 0

    while (top <= bottom && left <= right) {
      direction match {
        case 0 =>
          for (i <- left to right) {
            ans += matrix(top)(i)
          }
          top += 1
        case 1 =>
          for (i <- top to bottom) {
            ans += matrix(i)(right)
          }
          right -= 1
        case 2 =>
          for (i <- right to left by -1) {
            ans += matrix(bottom)(i)
          }
          bottom -= 1
        case 3 =>
          for (i <- bottom to top by -1) {
            ans += matrix(i)(left)
          }
          left += 1
      }
      direction = (direction + 1) % 4
    }
    ans.toList
  }
}

File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        int top = 0, left=0;
        int bottom = matrix.length-1, right = matrix[0].length-1;
        int direction = 0;

        while(top<=bottom && left<=right){
            //for col(left- right)
            if(direction ==0){
            for(int i =left; i<=right; i++){
                ans.add(matrix[top][i]);
            }
            ++top;
            }
            // for row downward
            if(direction ==1){
            for(int i=top; i<=bottom; i++){
            ans.add(matrix[i][right]);
            }
            --right;
            }
            // for col(right to left)
            if(direction ==2){
            for(int i=right; i>=left; i--){
                ans.add(matrix[bottom][i]);
            }
            --bottom;
            }
            //for row upwards
            if(direction ==3){
            for(int i=bottom; i>=top; i--){
                ans.add(matrix[i][left]);
            }
            ++left;
            }
            direction = (direction+1)%4;
        }
        return ans;
        
    }
}